src/functional/ffront/decorator.py

```diff
@@ -0,0 +1,227 @@
+                raise NotImplementedError("Only function closure vars are allowed currently.")
+
+        vars_ = ChainMap(self.closure_refs.globals, self.closure_refs.nonlocals)
+        if not all(func in vars_ for func in func_names):
+            undef_funcs = [func not in vars_ for func in func_names]
+            raise RuntimeError(f"Reference to undefined symbol(s) `{', '.join(undef_funcs)}`")
+        funcs_lowered = [vars_[func_names].__gt_itir__() for func_names in func_names]
+
+        return itir.Program(
+            function_definitions=funcs_lowered, fencil_definitions=[fencil_itir_node], setqs=[]
+        )
+
+    def _validate_args(self, *args, **kwargs):
+        # validate signature
+        if len(args) != len(self.past_node.params):
+            raise GTTypeError(
+                f"Function takes {len(self.past_node.params)} arguments, but {len(args)} were given."
+            )
+        if kwargs:
+            raise NotImplementedError("Keyword arguments are not supported yet.")
+
+    def __call__(self, *args, offset_provider, **kwargs):
+        self._validate_args(*args, **kwargs)
+
+        # extract size of all field arguments
+        size_args = []
+        for param_idx, param in enumerate(self.past_node.params):
+            if not isinstance(param.type, ct.FieldType):
+                continue
+            for dim_idx in range(0, len(param.type.dims)):
+                size_args.append(args[param_idx].shape[dim_idx])
+
+        backend = self.backend if self.backend else default_backend
+
+        execute_program(
+            self.itir, *args, *size_args, **kwargs, offset_provider=offset_provider, backend=backend
+        )
+
+
+def program(
+    definition: types.FunctionType, externals: Optional[dict] = None, backend: Optional[str] = None
+) -> Program:
+    return Program.from_function(definition, externals, backend)
+
+
+@dataclasses.dataclass(frozen=True)
+class FieldOperator:
+    foast_node: foast.FieldOperator
+    closure_refs: ClosureRefs
+    externals: dict[str, Any]
+    backend: str  # note: backend is only used if directly called
+    definition: Optional[types.FunctionType] = None
+
+    @classmethod
+    def from_function(
+        cls, definition: types.FunctionType, externals: Optional[dict] = None, backend: str = None
+    ):
+        closure_refs = ClosureRefs.from_function(definition)
+        foast_node = FieldOperatorParser.apply_to_function(definition)
+        return FieldOperator(
+            foast_node=foast_node,
+            closure_refs=closure_refs,
+            externals={} if externals is None else externals,
+            backend=backend,
+            definition=definition,
+        )
+
+    def __gt_type__(self) -> ct.FunctionType:
+        type_ = symbol_makers.make_symbol_type_from_value(self.definition)
+        assert isinstance(type_, ct.FunctionType)
+        return type_
+
+    def __gt_itir__(self):
+        return FieldOperatorLowering.apply(self.foast_node)
```

Comment:
I should think these two deserve docstrings. Ideally referring to an explanation of how they can be used on custom classes to incorporate any IteratorIR snippet elsewhere.

Reply:
There is a Protocol `GTCallable` now with some explanations. Examples on how to really use it will be added later when the concept matured. As you might have seen in the raw iterator decorator PR, where signature and definition don't play nice together, there is still room for improvement